File: haskell-exercises/fraction/src/Fraction.hs
module Fraction (Fraction, add, sub, mul, divide, hcf) where

type Fraction = (Int, Int)

simplify :: Fraction -> Fraction
simplify (n, d) = (n `div` h, d `div` h)
  where h = hcf n d

add :: Fraction -> Fraction -> Fraction
add n d = simplify $
  if den1 == den2 then (num1 + num2, den1)
  else (num1 * den2 + num2 * den1, den1 * den2)
  where
    (num1, den1) = n
    (num2, den2) = d

sub :: Fraction -> Fraction -> Fraction
sub n d = simplify $
  if den1 == den2 then (num1 - num2, den1)
  else (num1 * den2 - num2 * den1, den1 * den2)
  where
    (num1, den1) = n
    (num2, den2) = d

mul :: Fraction -> Fraction -> Fraction
mul n d = simplify $ (num1 * num2, den1 * den2)
  where
    (num1, den1) = n
    (num2, den2) = d

divide :: Fraction -> Fraction -> Fraction
divide n d = simplify $ (num1 * den2, den1 * num2)
  where
    (num1, den1) = n
    (num2, den2) = d

hcf :: Int -> Int -> Int
hcf n d = if d == 0 then n else hcf d (n `mod` d)

File: rust-exercises/fraction/tests/fraction.rs
use fraction::*;

#[test]
fn test_add() {
    let result = add((1, 2), (1, 4));
    assert_eq!(result, simplify(6, 8));
}

#[test]
fn test_sub() {
    let result = sub((1, 2), (1, 4));
    assert_eq!(result, simplify(2, 8));
}

#[test]
fn test_mul() {
    let result = mul((1, 2), (1, 4));
    assert_eq!(result, simplify(1, 8));
}

#[test]
fn test_divide() {
    let result = divide((1, 2), (1, 4));
    assert_eq!(result, simplify(4, 2));
}

#[test]
fn test_hcf_24_30() {
    let result = hcf(24, 30);
    assert_eq!(result, 6)
}

#[test]
fn test_hcf_30_24() {
    let result = hcf(30, 24);
    assert_eq!(result, 6)
}

#[ignore]
#[test]
fn test_add_simplifying() {
    let result = add((1, 2), (1, 4));
    assert_eq!(result, simplify(3, 4));
}
